### ISSUE TITLE
Qwen Edit very small difference in time?

### ISSUE THREAD

**Issue by alexpuliatti** [fal-ai/flashpack#12]
Made a small benchmark on my finetune of QE 2509 and the results on a H100 were: 
Normal loading: 79.5 sec
Flashpack: 76 sec 

Wondering if these are similar to your results? I thought it would have been a bit more a significant speed up.

**Comment by painebenjamin** [comment]
Hello @alexpuliatti, thanks for checking this out!

Both of those numbers are very high for H100 if the weights are disk cached, but could be normal if they're not. Here's my test setup on QE 2509.

## Hardware

Single H100 HBM3 node, Ubuntu 22.04, CUDA 12.4, 16-core Xeon Platinum 8468, fairly slow disk:

```
sudo hdparm -Tt /dev/vda

/dev/vda:
 Timing cached reads:   23572 MB in  2.00 seconds = 11808.18 MB/sec
 Timing buffered disk reads: 1774 MB in  3.00 seconds = 590.45 MB/sec
```

## Load Times

| Method | Load Time (Seconds, Uncached) | Load Time (Seconds, Cached) |
| -------- | ------------------- | ---------------- |
| Accelerate | 135 | 10.5 |
| FlashPack | 135 | 3.7 |

> Why are the uncached times identical?

FlashPack loads the weights from disk faster, but the weights are almost the exact same size in terms of raw bytes. In this case, the weights are roughly 54 gigabytes, meaning if we average ~500 MB/sec it will take at least 108 seconds to read that data from the underlying disk. In such cases, the advantage that FlashPack can provide is very slim, as the vast majority of the effort needs to occur before Accelerate or FlashPack would even be able to access the weights.

> Can I count on weights being cached?

This answer will vary significantly depending on your hardware and software. You can be sure that it will be cached right after downloading it and right after reading it once, but other than that, it depends on when your machine was last shut down, what the cache eviction policy is, the kind of disk you're reading from, and any additional ephemeral or persistent caches you have attached to the machine.

## Benchmark Code

I'm working on making it easier to transform entire pipelines at once so you don't need quite so much boilerplate to do it.

```py
from diffusers import QwenImageEditPlusPipeline
from huggingface_hub import snapshot_download
from PIL import Image
import requests
import torch
import time
import gc
import os

from transformers import Qwen2_5_VLForConditionalGeneration, Qwen2Tokenizer, Qwen2VLProcessor
from diffusers.models import QwenImageTransformer2DModel, AutoencoderKLQwenImage
from diffusers.schedulers import FlowMatchEulerDiscreteScheduler
from flashpack.integrations.diffusers import FlashPackDiffusionPipeline, FlashPackDiffusersModelMixin
from flashpack.integrations.transformers import FlashPackTransformersModelMixin
from flashpack.utils import timer

# Checkpoint and FlashPack directories
checkpoint_dir = snapshot_download("Qwen/Qwen-Image-Edit-2509")
flashpack_dir = "Qwen-Image-Edit-2509-FlashPack"
skip_original = os.getenv("SKIP_ORIGINAL", "0") == "1"

image_1_url = "https://v3.fal.media/files/monkey/i3saq4bAPXSIl08nZtq9P_ec535747aefc4e31943136a6d8587075.png"
image_2_url = "https://v3.fal.media/files/penguin/BCOZp6teRhSQFuOXpbBOa_da8ef9b4982347a2a62a516b737d4f21.png"
image_3_url = "https://v3.fal.media/files/tiger/sCoZhBksx9DvwSR4_U3_C_3d1f581441874005908addeae9c10d0f.png"

image1 = Image.open(requests.get(image_1_url, stream=True).raw)
image2 = Image.open(requests.get(image_2_url, stream=True).raw)
image3 = Image.open(requests.get(image_3_url, stream=True).raw)
generator = torch.Generator("cuda").manual_seed(0)

inputs = {
    "image": [image1, image2, image3],
    "prompt": "Close shot of a woman standing in next to this car on this highway",
    "generator": generator,
    "true_cfg_scale": 4.0,
    "negative_prompt": " ",
    "num_inference_steps": 40,
    "guidance_scale": 1.0,
    "num_images_per_prompt": 1,
}

# FlashPack classes

class Qwen2_5_VLForConditionalGenerationFlashPack(FlashPackTransformersModelMixin, Qwen2_5_VLForConditionalGeneration):
    pass

class QwenImageTransformer2DModelFlashPack(FlashPackDiffusersModelMixin, QwenImageTransformer2DModel):
    pass

class AutoencoderKLQwenImageFlashPack(FlashPackDiffusersModelMixin, AutoencoderKLQwenImage):
    pass

class QwenImageEditPlusPipelineFlashPack(FlashPackDiffusionPipeline, QwenImageEditPlusPipeline):
    def __init__(
        self,
        transformer: QwenImageTransformer2DModelFlashPack,
        vae: AutoencoderKLQwenImageFlashPack,
        tokenizer: Qwen2Tokenizer,
        text_encoder: Qwen2_5_VLForConditionalGenerationFlashPack,
        scheduler: FlowMatchEulerDiscreteScheduler,
        processor: Qwen2VLProcessor,
    ) -> None:
        super().__init__(
            transformer=transformer,
            vae=vae,
            tokenizer=tokenizer,
            text_encoder=text_encoder,
            scheduler=scheduler,
            processor=processor,
        )

if not skip_original:
    with timer("individual from_pretrained load"):
        transformer = QwenImageTransformer2DModelFlashPack.from_pretrained(
            os.path.join(checkpoint_dir, "transformer"),
            torch_dtype=torch.bfloat16,
            device_map="cuda",
        )
        vae = AutoencoderKLQwenImageFlashPack.from_pretrained(
            os.path.join(checkpoint_dir, "vae"),
            torch_dtype=torch.bfloat16,
            device_map="cuda",
        )
        text_encoder = Qwen2_5_VLForConditionalGenerationFlashPack.from_pretrained(
            os.path.join(checkpoint_dir, "text_encoder"),
            torch_dtype=torch.bfloat16,
            device_map="cuda",
        )
        tokenizer = Qwen2Tokenizer.from_pretrained(
            os.path.join(checkpoint_dir, "tokenizer"),
        )
        scheduler = FlowMatchEulerDiscreteScheduler.from_pretrained(
            os.path.join(checkpoint_dir, "scheduler"),
        )
        processor = Qwen2VLProcessor.from_pretrained(
            os.path.join(checkpoint_dir, "processor"),
        )
        pipeline = QwenImageEditPlusPipelineFlashPack(
            transformer=transformer,
            vae=vae,
            tokenizer=tokenizer,
            processor=processor,
            text_encoder=text_encoder,
            scheduler=scheduler,
        )

    # Run an inference to catch lazy load time differences
    with timer("inference"):
        with torch.inference_mode():
            pipeline(**inputs)

    # Save FlashPack
    if not os.path.exists(flashpack_dir):
        with timer("saving FlashPack"):
            pipeline.save_pretrained_flashpack(flashpack_dir, silent=False)

    # Delete original pipeline and clear memory
    del pipeline
    del transformer
    del vae
    del text_encoder
    del tokenizer
    del scheduler
    del processor
    torch.cuda.empty_cache()
    torch.cuda.synchronize()
    torch.cuda.ipc_collect()
    gc.collect()
else:
    assert os.path.exists(flashpack_dir), "FlashPack directory does not exist, cannot skip original. Re-run with SKIP_ORIGINAL=0"


# Load pipeline from FlashPack
with timer("loading FlashPack"):
    pipeline = QwenImageEditPlusPipelineFlashPack.from_pretrained_flashpack(flashpack_dir, device_map="cuda")

# Run an inference to catch any lazy load time differences (there are none, but let's prove it)
generator.manual_seed(0)
with timer("FlashPack inference"):
    with torch.inference_mode():
        pipeline(**inputs)
```

And my console output:

```
$ python qwen.py
Fetching 35 files: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 35/35 [00:00<00:00, 251802.13it/s]
Loading checkpoint shards: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 5/5 [00:06<00:00,  1.29s/it]
`torch_dtype` is deprecated! Use `dtype` instead!
Loading checkpoint shards: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 4/4 [00:02<00:00,  1.48it/s]
individual from_pretrained load: 10.45s
guidance_scale is passed as 1.0, but ignored since the model is not guidance-distilled.
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 40/40 [01:36<00:00,  2.42s/it]
inference: 1m38s
Loading pipeline components...: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 6/6 [00:03<00:00,  1.61it/s]
loading FlashPack: 3.74s
guidance_scale is passed as 1.0, but ignored since the model is not guidance-distilled.
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 40/40 [01:36<00:00,  2.42s/it]
FlashPack inference: 1m38s
```

You can run it again with `SKIP_ORIGINAL=1 python qwen.py` to skip loading the original pipeline to be certain it's not cached in-process. To clear filesystem cache on Linux, run `sudo sync; echo 3 | sudo tee /proc/sys/vm/drop_caches`.
